[2단계 - 연관 관계 매핑] 리뷰요청 드립니다~!!

byunghakjang1230 opened the pull request:
연관관계 설정을 하다보니 서비스 클래스와 서비스 테스트 코드까지 수정을 해야했는데 맞게 진행한거겠죠??🙀

연관관계 설정 이후 해당 연관관계로 설정 된 객체로 대상 목록을 조회하는 부분에서 조금 헷갈렸습니다. 단방향만 설정한 상태에서 메소드 이름 기준으로 쿼리를 만들어 내는 부분에서 계속 막히더군요. 양방향 설정을 하고 메소드에서 컬럼 명을 객체 변수명으로 수정하고(WriterId -> Writer), 매개변수로 객체를 던지니 원하던 결과로 조회되는걸 확인하였습니다. 
그렇게 진행하다 보니 서비스에서 사용되는 Repository 메소드 명을 전부 수정해야하는 상황이 되어서...ㅎㅎㅎ
그렇게 수정을 하다보니 Mock에서 발생하는 에러는 접하게 되었고 그 원인을 찾는데 조금 어려움이 있었지만 Mock에 대해서도 자연스럽게 학습을 하게 되어서 좋았던 시간이었습니다.
 
놓친 부분이나 부족한 부분들에 대한 코멘트 부탁드립니다!🙏

ps. 아, 궁금한게 있습니다.
Mock을 학습하다보니 DB와 관련된 테스트를 진행하기에 매우 편리해 보이긴 했는데 Repository에서 Mock을 사용하는게 의미가 있는건지 궁금해졌습니다. 제가 느끼기에는 메소드의 결과값을 세팅해두고(정답을 담아두고..) 그렇게 반환된 결과와 결과값을 비교하는게 되는 모양이 되더라구요. 제가 테스트 방법을 잘못한건지...😓 저런식의 테스트는 의미가 없어보이는데 Repository에서도 Mock을 사용하여 테스트를 하나요? 한다면...제가 생각하지 못한 다른 의미가 있는 것일까요??😲

hilu0318 committed:
refactor(User.java): Question 클래스에 대한 연관관계 추가

refactor(Question.java): User 클래스에 대한 연관관계 추가

refactor(QuestionRepository.java): User 클래스에 대한 연관관계 추가로 인한 메소드 명, 매개변수 타입 수정

refactor(QuestionRepositoryTest.java): QuestionRepository의 메소드 수정으로 인한 관련 테스트 코드 수정

hilu0318 committed:
refactor(User.java): Answer 클래스에 대한 연관관계 추가

refactor(Question.java): 기본 생성자 추가

refactor(Answer.java): 기본 생성자 및 User, Question 클래스의 연관관계 추가

refactor(AnswerRepository.java): Answer의 연관관계 추가에 따른 메소드 이름 및 매개변수 타입 변경

refactor(AnswerRepositoryTest.java): AnswerRepository 메서드 변경에 따른 관련 코드 수정

hilu0318 committed:
refactor(User.java): DeleteHistory 클래스에 대한 연관관계 추가

refactor(DeleteHistory.java): User 클래스에 대한 연관관계 추가

refactor(DeleteHistoryRepository.java): DeleteHistory 연관관계 설정으로 메소드 이름 및 매개변수 타입 수정

refactor(DeleteHistoryRepositoryTest.java): 연관관계 설정에 의한 메소드 수정으로 관련 테스트코드 수정

hilu0318 committed:
refactor(Answer.java): 연관관계의 주인 클래스에 객체 연결 설정

refactor(AnswerTest.java): 연관관계 설정에 의한 매소드 이름 변경

refactor(QuestionTest.java): 연관관계 설정에 의한 매소드 이름 변경

refactor(User.java): 불필요한 setter 메소드 제거

refactor(UserRepositoryTest.java): setter 메소드 제거로 관련 테스트 코드 수정

hilu0318 committed:
refactor(DeleteHistory.java): 생성자 매개변수 타입 변경

- Long deletedById -> User deletedByUser로 변경

refactor(QnaService.java): 도메인 연관관계 설정으로 인해 Repository들의 수정된 메소드 이름으로 수정

refactor(QnaServiceTest.java): QnaService 클래스 및 Repository의 변경사항 테스트 코드에 반영

hilu0318 committed:
docs(README.md): 학습내용 정리

hilu0318 committed:
refactor(BaseEntity.java): 불필요한 setter 제거

hilu0318 committed:
refactor(Answer.java): 사용하지 않는 메서드 삭제 및 연관관계 값 주입 코드 추가

refactor(Question.java): Answer 클래스에 대한 양방향 연관관계 설정 및 사용하지 않는 코드 삭제

young891221 commented:
> ps. 아, 궁금한게 있습니다.
> Mock을 학습하다보니 DB와 관련된 테스트를 진행하기에 매우 편리해 보이긴 했는데 Repository에서 Mock을 사용하는게 의미가 있는건지 궁금해졌습니다. 제가 느끼기에는 메소드의 결과값을 세팅해두고(정답을 담아두고..) 그렇게 반환된 결과와 결과값을 비교하는게 되는 모양이 되더라구요. 제가 테스트 방법을 잘못한건지...😓 저런식의 테스트는 의미가 없어보이는데 Repository에서도 Mock을 사용하여 테스트를 하나요? 한다면...제가 생각하지 못한 다른 의미가 있는 것일까요??😲

저도 같은 생각입니다ㅎㅎ
실제로 사용한다면 빠른 테스트를 위해 DB의존성을 없애고 원하는 형식이 나오는지 여부에 대한 repository Mock 테스트가 있을 수도 있지만, 불필요하다 생각되네요. 실제 DB에서 올바른 값이 나오는지에 대한 테스트도 무조건 있어야 하기 때문에 오히려 테스트가 중복될 수도 있다고 생각합니다.

young891221 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -15,18 +24,23 @@
 
-    @Column(name = "writer_id")
-    private Long writerId;
+    @ManyToOne
```
ManyToOne은 Eager 조회가 디폴트입니다. JPA를 사용한다면 메리트가 있는 LAZY로 변경해 주시는게 어떨까요?ㅎ
(다른 부분들도 모두 부탁드립니다)

young891221 replied:
LAZY 변경 후 해당하는 동작이 잘 되는지 테스트 코드로 확인해 주세요!

byunghakjang1230 replied:
이게 테스트를 하기가 좀 힘들었네요.ㅜㅜ repository를 통해서 등록을 하면 전부 영속성에 포함되어서 지연로딩을 구현할 수 없어서 별도의 sql파일로 지연로딩 테스트용 데이터만 따로 등록해서 테스트를 진행했습니다. 
각각의 RepositoryTest 클래스에 별도의 지연로딩 테스트 케이스를 작성해 두었습니다~
(혹시나 다른 테스트 방법이 있을까요..?? 한참을 찾아도 repository를 사용해서는 영속성 제거하는 방법을 찾을수가 없었네요..ㅜㅠ)

young891221 replied:
지금 하신 정도로만 하셔도 괜찮을것 같습니다ㅎ 잘 되지만 연습삼아 처리해 보는 형태라서요!

young891221 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -15,18 +24,23 @@
     @Column(nullable = false)
     private boolean deleted = false;
 
+    public Answer() {
```
디폴트 생성자는 함부로 사용하면 안되기에 private or protected로 변경해 주세요!

byunghakjang1230 replied:
넵, 전부 protected로 선언해두었습니다~

young891221 commented on src/main/java/qna/domain/AnswerRepository.java:
```diff
@@ -6,13 +6,13 @@
+    List<Answer> findByQuestionIsNotNull();
 
-    List<Answer> findByWriterIdAndQuestionId(Long writerId, Long questionId);
+    List<Answer> findByWriterAndQuestion(User writer, Question question);
```
writerId를 사용하면 조회없이 ID값으로 가능한데 writer는 writerId를 사용하여 디비를 한번 조회하고 결과값을 사용해야 해서 처음의 형식을 유지해 주는게 더 좋아보입니다ㅎ

young891221 replied:
repository에서 ID사용이 변경된 부분들은 아무래도 잘못된것 같습니다ㅎ
맵핑은 올바르게 된거 같은데...다시 한번 확인 부탁드립니다:)

byunghakjang1230 replied:
전체적으로 다시 코드를 수정하였습니다~ 근데 제가 임의로 만든 Repository 메소드 명이 문제인지 findBydeleteById로 메서드명을 만드니 에러가 발생해서 학습용으로 만들어본 메서드라 우선 삭제처리 하고 추후에 필요하면 다시 생성해보도록 하겠습니다~

young891221 reviewed (changes requested):
지난번에 이어서 이번 미션 구현도 잘 해주셨네요 💯 
질문에 대한 답변과 함께 몇 가지 코멘트 드렸습니다.
확인 부탁드려요~!

hilu0318 committed:
feat(lazy_data.sql): 지연로딩 테스트를 위해 영속성에 포함되지 않을 데이터 입력 쿼리 작성

build(application.properties): 지연로딩 테스트 데이터 입력쿼리 파일 위치 지정.

refactor(User.java): 생성자 접근 제한자 protected로 수정

feat(UserRepositoryTest.java): 지연로딩 테스트 코드 추가

refactor(UserRepositoryTest.java): 일부 테스트케이스 삭제 및 수정

- 지연로딩 테스트 데이터 주입으로 일부 데이터가 영속성에 포함되지 않기에 영향도 있는 테스트 코드 삭제 및 수정

refactor(Question.java): 연관관계 설정에서 지연로딩 옵션(LAZY) 지정

- 생성자의 접근 제어자를 protected로 수정

refactor(QuestionRepository.java): 매개변수를 객체에서 Long 타입으로 수정

test(QuestionRepositoryTest.java): 지연로딩 테스트코드 작성

refactor(QuestionRepositoryTest.java): Repository의 수정된 메소드로 관련코드 수정

refactor(Answer.java): 연관관계 설정에서 지연로딩 옵션(LAZY) 지정

refactor(AnswerRepository.java): 매개변수를 객체에서 Long 타입으로 수정

feat(AnswerRepositoryTest.java): 지연로딩 테스트 코드 작성

refactor(AnswerRepositoryTest.java): Repository 메소트 수정에 따른 테스트 코드 수정

feat(DeleteHistory.java): protected 생성자 추가 및 deletedByUser의 getter 생성

feat(DeleteHistoryRepositoryTest.java): 지연로딩 테스트코드 작성

refactor(QnaService.java): 코드에 사용된 객체들의 메소트 변경으로 관련 코드 수정

refactor(QnaServiceTest.java): 코드에 사용된 객체들의 메소트 변경으로 관련 코드 수정

hilu0318 committed:
refactor(DeleteHistoryRepository.java): 학습용 생성 메서드 삭제

feat(UserRepositoryTest.java): 삭제된 메서드의 테스트 케이스 제거

young891221 reviewed (approved):
리뷰 반영 잘 해주셨네요 👍 
추가로 드릴 코멘트가 없습니다. 다음 단계 진행해 주세요!
